samples/kinesis_file_gstreamer_sample_app.cpp

```diff
@@ -0,0 +1,797 @@
+#define DEFAULT_ABSOLUTE_FRAGMENT_TIMES TRUE
+#define DEFAULT_FRAGMENT_ACKS TRUE
+#define DEFAULT_RESTART_ON_ERROR TRUE
+#define DEFAULT_RECALCULATE_METRICS TRUE
+#define DEFAULT_STREAM_FRAMERATE 25
+#define DEFAULT_AVG_BANDWIDTH_BPS (4 * 1024 * 1024)
+#define DEFAULT_BUFFER_DURATION_SECONDS 120
+#define DEFAULT_REPLAY_DURATION_SECONDS 40
+#define DEFAULT_CONNECTION_STALENESS_SECONDS 60
+#define DEFAULT_CODEC_ID "V_MPEG4/ISO/AVC"
+#define DEFAULT_TRACKNAME "kinesis_video"
+#define DEFAULT_FRAME_DURATION_MS 1
+#define DEFAULT_CREDENTIAL_ROTATION_SECONDS 3600
+#define DEFAULT_CREDENTIAL_EXPIRATION_SECONDS 180
+#define DEFAULT_TOTAL_TRACK_COUNT 2
+#define DEFAULT_AUDIO_VIDEO_DRIFT_TIMEOUT_SECOND 5
+
+#define DEFAULT_VIDEO_TRACKID 1
+#define DEFAULT_AUDIO_TRACK_NAME "audio"
+#define DEFAULT_AUDIO_CODEC_ID "A_AAC"
+#define DEFAULT_AUDIO_TRACKID 2
+
+typedef struct _FileInfo {
+    _FileInfo():
+            path(""),
+            last_fragment_ts(0) {}
+    string path;
+    uint64_t last_fragment_ts;
+    mutex mtx;
+} FileInfo;
+
+typedef struct _CustomData {
+
+    _CustomData():
+            first_video_frame(true),
+            eos_triggered(false),
+            stream_status(STATUS_SUCCESS),
+            key_frame_pts(0),
+            main_loop(NULL) {
+        producer_start_time = chrono::duration_cast<nanoseconds>(systemCurrentTime().time_since_epoch()).count();
+    }
+
+    GMainLoop *main_loop;
+    // queue elements to be linked to demuxer's sometimes pad in demux_pad_cb
+    GstElement *video_queue, *audio_queue;
+    GstElement *pipeline;
+
+    unique_ptr<KinesisVideoProducer> kinesis_video_producer;
+    shared_ptr<KinesisVideoStream> kinesis_video_stream;
+    char *stream_name;
+
+    mutex audio_video_sync_mtx;
+    condition_variable audio_video_sync_cv;
+
+    // file to upload
+    FileInfo file;
+
+    // indicate whether a video key frame has been received or not.
+    volatile bool first_video_frame;
+
+    // when uploading file, whether one of audio or video pipeline has reached eos.
+    atomic_bool eos_triggered;
+
+    // key:     trackId
+    // value:   whether application has received the first frame for trackId.
+    map<int, bool> stream_started;
+
+    // stores any error status code reported by StreamErrorCallback.
+    atomic_uint stream_status;
+
+    // When uploading file, store the pts of frames that has flag FRAME_FLAG_KEY_FRAME. When the entire file has been uploaded,
+    // key_frame_pts contains the timetamp of the last fragment in the file. key_frame_pts is then stored into last_fragment_ts
+    // of the file.
+    // Unit: ns
+    uint64_t key_frame_pts;
+
+    // Assuming frame timestamp are relative. Add producer_start_time to each frame's
+    // timestamp to convert them to absolute timestamp. This way fragments dont overlap after token rotation when doing
+    // file uploading.
+    uint64_t producer_start_time;
+
+    unique_ptr<Credentials> credential;
+} CustomData;
```

Comment:
BTW, why not moving this into the namespace

Reply:
IMO, I think it's not appropriate since this is an application scope, not the library scope.

But, I honestly don't have a hard feeling either way, please let me know what you think.